gamemodes/terrortown/gamemode/roles/taskmaster/taskmaster.lua

```diff
@@ -0,0 +1,149 @@
+        end
+    end
+    -- TODO: Handle edge case where there are no valid tasks. Maybe we need to change their role? Or just have a free fallback task?
+    return false
+end
+
+function plymeta:RemoveTask(taskId)
+    if not self:IsActiveTaskmaster() then return end
+
+    local isKillTask = TASKMASTER.killTasks[taskId] and true or false
+    local taskList = isKillTask and TASKMASTER.killTasks or TASKMASTER.miscTasks
+    local activeTasksName = isKillTask and "taskmasterKillTasks" or "taskmasterMiscTasks"
+
+    taskList[taskId].OnTaskRemoved(self)
+
+    table.RemoveByValue(self[activeTasksName], taskId)
+    self:SetProperty(activeTasksName, self[activeTasksName], self)
+end
+
+function plymeta:RerollTask(taskId)
+    if not self:IsActiveTaskmaster() then return end
+
+    local isKillTask = TASKMASTER.killTasks[taskId] and true or false
+    local activeTasksName = isKillTask and "taskmasterKillTasks" or "taskmasterMiscTasks"
+    local index = table.KeyFromValue(self[activeTasksName], taskId)
+    if not index then return end
+
+    self:AssignTask(isKillTask, index)
+    self:RemoveTask(taskId)
+end
+
+function plymeta:CompleteTask(taskId)
+    if not self:IsActiveTaskmaster() then return end
+
+    local isKillTask = TASKMASTER.killTasks[taskId] and true or false
+    local taskList = isKillTask and TASKMASTER.killTasks or TASKMASTER.miscTasks
+    local activeTasksName = isKillTask and "taskmasterKillTasks" or "taskmasterMiscTasks"
+    if table.HasValue(self[activeTasksName], taskId) then
+        taskList[taskId].OnTaskComplete(self)
+        table.insert(self.taskmasterCompletedTasks, taskId)
+        self:SetProperty("taskmasterCompletedTasks", self.taskmasterCompletedTasks, self)
+
+        local activeTasksList = table.Copy(self.taskmasterKillTasks)
+        table.Add(activeTasksList, self.taskmasterMiscTasks)
+        for _, id in ipairs(self.taskmasterCompletedTasks) do
+            if table.HasValue(activeTasksList, id) then
+                table.RemoveByValue(activeTasksList, id)
+            else
+                break
+            end
+        end
+        if #activeTasksList == 0 then
+            self:SetProperty("taskmasterShouldWin", true)
+        end
+        return true
+    end
+    return false
+end
+
+ROLE_ON_ROLE_ASSIGNED[ROLE_TASKMASTER] = function(ply)
+    for _ = 1, taskmaster_kill_tasks:GetInt() do
+        ply:AssignTask(true)
+    end
+    for _ = 1, taskmaster_misc_tasks:GetInt() do
+        ply:AssignTask(false)
+    end
+end
+
+------------------
+-- WIN BLOCKING --
+------------------
+
+hook.Add("TTTWinCheckBlocks", "Taskmaster_TTTWinCheckBlocks", function(win_blocks)
+    table.insert(win_blocks, function(win_type)
+        if win_type == WIN_NONE then return win_type end
+
+        local taskmaster = player.GetLivingRole(ROLE_TASKMASTER)
+        if not IsPlayer(taskmaster) then return win_type end
+
+        if taskmaster.taskmasterShouldWin then return win_type end
```

Comment:
Shouldn't this be it's own win type?
```suggestion
        if taskmaster.taskmasterShouldWin then return WIN_TASKMASTER end
```

Reply:
I think the taskmaster winning should be a secondary win condition and not a primary one

Reply:
It's a secondary win but it blocks primary wins until they complete?